src/machine-learning/machine-learning-api-client.ts

```diff
@@ -0,0 +1,195 @@
+/*!
+ * Copyright 2020 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { HttpRequestConfig, HttpClient, HttpError, AuthorizedHttpClient } from '../utils/api-request';
+import { PrefixedFirebaseError } from '../utils/error';
+import { FirebaseMachineLearningError, MachineLearningErrorCode } from './machine-learning-utils';
+import * as utils from '../utils/index';
+import * as validator from '../utils/validator';
+import { FirebaseApp } from '../firebase-app';
+
+const ML_V1BETA1_API = 'https://mlkit.googleapis.com/v1beta1';
+const FIREBASE_VERSION_HEADER = {
+  'X-Firebase-Client': 'fire-admin-node/<XXX_SDK_VERSION_XXX>',
+};
+
+export interface StatusErrorResponse {
+    readonly code: number;
+    readonly message: string;
+}
+
+export interface ModelContent {
+  readonly displayName?: string;
+  readonly tags?: string[];
+  readonly state?: {
+    readonly validationError?: StatusErrorResponse;
+    readonly published?: boolean;
+  };
+  readonly tfliteModel?: {
+    readonly gcsTfliteUri: string;
+    readonly sizeBytes: number;
+  };
+}
+
+export interface ModelResponse extends ModelContent {
+  readonly name: string;
+  readonly createTime: string;
+  readonly updateTime: string;
+  readonly etag: string;
+  readonly modelHash?: string;
+}
+
+
+/**
+ * Class that facilitates sending requests to the Firebase ML backend API.
+ *
+ * @private
+ */
+export class MachineLearningApiClient {
+  private readonly httpClient: HttpClient;
+  private projectIdPrefix?: string;
+
+  constructor(private readonly app: FirebaseApp) {
+    if (!validator.isNonNullObject(app) || !('options' in app)) {
+      throw new FirebaseMachineLearningError(
+        'invalid-argument',
+        'First argument passed to admin.machineLearning() must be a valid '
+          + 'Firebase app instance.');
+    }
+
+    this.httpClient = new AuthorizedHttpClient(app);
+  }
+
+  public getModel(modelId: string): Promise<ModelResponse> {
+    return Promise.resolve()
+      .then(() => {
+        return this.getModelName(modelId);
+      })
+      .then((modelName) => {
+        return this.getResource<ModelResponse>(modelName);
+      });
+  }
+
+
+  private getUrl(): Promise<string> {
```

Comment:
Nit: I'd recommend pushing helpers like this to the bottom of the class, so that more important functionality like `getResouce` appears first.

Reply:
done